### PR TITLE
add channel last 3d support for batch_norm on CPU

### PR DESCRIPTION
**Pull request by CaoE**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98819
* #97775
* __->__ #97774

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: test/test_nn.py [hunk 1]
```diff
@@ -4945,15 +4946,15 @@ def helper(self, size, dtype, mixed_dtype=False):
...
+            self.assertTrue(out.is_contiguous(memory_format=t_type))
             self.assertTrue(ref_out.is_contiguous())
             self.assertEqual(out, ref_out)
             self.assertEqual(bn.weight.grad, ref_bn.weight.grad)
             self.assertEqual(bn.bias.grad, ref_bn.bias.grad)
             self.assertEqual(input.grad, ref_input.grad)
 
         # test NC11 and N1HW; test mixed dtype
-        for shape in [(4, 8, 10, 10), (4, 1, 9, 9), (4, 9, 1, 1)]:
+        for shape in [(4, 8, 10, 10), (4, 1, 9, 9), (4, 9, 1, 1), (4, 8, 2, 10, 10), (4, 1, 2, 9, 9), (4, 9, 1, 1, 1)]:
```
Comment (mingfeima) [inline test/test_nn.py:4957]: add an argument of `mod` is more neat:
```
helper(self, nn.BatchNorm2d, shape, ...)
helper(self, nn.BatchNorm3d, shape, ...)
```

**Review by mingfeima — APPROVED** [review]
Generally LGTM, just need to update the test cases a little bit.

File: test/test_nn.py (hunk 1 above)
Comment (CaoE) [reply test/test_nn.py]: Added an argument of mod.

**Review by jgong5 — APPROVED** [review]

**Comment by CaoE** [comment]
@ngimel Could you please review this PR ? Thank you.

**Comment by CaoE** [comment]
@ngimel Could you please review this PR ? Thank you.

**Comment by CaoE** [comment]
@malfet Could you please review this PR ? Thank you.

**Comment by CaoE** [comment]
@cpuhrsch Could you please review this PR ? Thank you.

**Comment by cpuhrsch** [comment]
@CaoE - Which PR introduced the but that this is fixing?

**Comment by CaoE** [comment]
> @CaoE - Which PR introduced the but that this is fixing?

@cpuhrsch This PR is not fixing specific PRs. Current BN kernel can actually support `ChannelsLast3d`, but `ChannelsLast3d` is not supported because the condition do not include `ChannelsLast3d`. The title of this PR may be a bit confusing, I have modified it.

**Comment by cpuhrsch** [comment]
Adding @mikaylagawarecki since this also touches torch.nn. 

@mikaylagawarecki - I think this looks good, but maybe you can double check the tests are sufficient?

File: test/test_nn.py [hunk 2]
```diff
@@ -5062,20 +5062,20 @@ def test_hardtanh_backward(self):
         self.assertEqual(x.grad, x_grad_ref)
 
     def test_batchnorm_nhwc_cpu(self):
-        def helper(self, size, dtype, mixed_dtype=False):
+        def helper(self, mod, size, dtype, mixed_dtype=False, format=torch.channels_last):
```
[the remark this answers is not in the document]
Comment (CaoE) [reply test/test_nn.py:5065]: > the effect it achieves is similar to that of this test except for
1. the functionality of the mixed_dtype arg
2. the checks on backward

Yes.

> BatchNorm3d (which has 5d sample inputs) does not seem to have supported channels_last_3d prior to this PR (I don't see the test being xfailed in the ModuleInfo)

Actually, BatchNorm3d can not utilize channels_last_3d kernel prior to this PR but it can has channels_last_3d support using another code path by TensorIterator https://github.com/pytorch/pytorch/pull/97774/files#diff-9e36531b6ea57776251e04cf2e1b84a94145020d870fed3d7f82388c47371483L177.

[the remark this answers is not in the document]
Comment (mikaylagawarecki) [reply test/test_nn.py:5065]: Ah I see, so this PR allows BatchNorm3d to use a fast kernel for `channels_last_3d` inputs without transforming the inputs, is that correct?

Would you be open to updating [test_memory_format](https://github.com/pytorch/pytorch/blob/main/test/test_modules.py#L604)? I don't want to block this PR if the change is urgent but it would be some super nice cleanup since it seems like you are updating other functions to support `channels_last_3d` as well and could catch cases where backward kernels are not properly supporting `channels_last_3d` :)

[the remark this answers is not in the document]
Comment (CaoE) [reply test/test_nn.py:5065]: Yes, I'm willing to add checks on backward for `test_memory_format`.

**Review by mikaylagawarecki — APPROVED** [review]
@CaoE awesome thank you! I will review the other PR. We can land this stack first, stamping to unblock
